CMakeLists.txt

```
@@ -0,0 +1,60 @@
+cmake_minimum_required(VERSION 3.14 FATAL_ERROR)
+
+project(
+  CPM
+  VERSION 0.37.0
+  LANGUAGES NONE
+)
+
+include(CMakePackageConfigHelpers)
+configure_package_config_file(
+  "${CMAKE_CURRENT_SOURCE_DIR}/cmake/CPMConfig.cmake.in"
+  "${CMAKE_CURRENT_BINARY_DIR}/CPMConfig.cmake"
+  INSTALL_DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/CPM"
+)
+write_basic_package_version_file(
+  "${CMAKE_CURRENT_BINARY_DIR}/CPMConfigVersion.cmake" COMPATIBILITY SameMajorVersion
+                                                                     ARCH_INDEPENDENT
+)
+
+# For consumer using FetchContent
+if(NOT ${CMAKE_PROJECT_NAME} STREQUAL ${PROJECT_NAME})
+  # Trick to use the find_package
+  configure_file(
+    "${CMAKE_CURRENT_SOURCE_DIR}/cmake/CPM.cmake" "${CMAKE_CURRENT_BINARY_DIR}/CPM.cmake" COPYONLY
+  )
+  set(CPM_DIR "${CMAKE_CURRENT_BINARY_DIR}")
+  # Check if the user download the repo with git and not in an official release
+  if(EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/.git" AND IS_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/.git")
+    find_package(Git REQUIRED QUIET)
+    # Generate a git-describe version string from Git repository tags.
+    execute_process(
+      COMMAND ${GIT_EXECUTABLE} tag --points-at HEAD
+      WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+      OUTPUT_VARIABLE CPM_TAG
+      RESULT_VARIABLE GIT_DESCRIBE_ERROR_CODE
+      OUTPUT_STRIP_TRAILING_WHITESPACE
+    )
+    if(NOT GIT_DESCRIBE_ERROR_CODE AND NOT CPM_TAG STREQUAL "")
+      set(CPM_RELEASE TRUE)
+    else()
+      set(CPM_RELEASE FALSE)
+    endif()
+  else()
+    set(CPM_RELEASE TRUE)
+  endif()
+  find_package(CPM CONFIG REQUIRED NO_POLICY_SCOPE)
```

We are over-complicating it here. This works just fine
```suggestion
  include(cmake/CPM.cmake)
```

---

Ok, I found a small bug with how we are looking for the .git I can change this and your changes.

I tested using : 

```cmake
 
cmake_minimum_required(VERSION 3.24)
project(TestCPMFetchContent)

include(FetchContent)
FetchContent_Declare(CPM
		GIT_REPOSITORY https://github.com/flagarde/CPM.cmake
		GIT_TAG FetchContent
		OVERRIDE_FIND_PACKAGE)


FetchContent_MakeAvailable(CPM)
find_package(CPM)
# the install option has to be explicitly set to allow installation
CPMAddPackage(
  GITHUB_REPOSITORY jarro2783/cxxopts
  VERSION 2.2.1
  OPTIONS "CXXOPTS_BUILD_EXAMPLES NO" "CXXOPTS_BUILD_TESTS NO" "CXXOPTS_ENABLE_INSTALL YES"
)
```

In this case not loop but a crash : 
```bash
CMake Error at CMakeLists.txt:15 (CPMAddPackage):
  Unknown CMake command "CPMAddPackage".


-- Configuring incomplete, errors occurred!
```

without  OVERRIDE_FIND_PACKAGE : 

```bash
-- The C compiler identification is GNU 12.2.1
-- The CXX compiler identification is GNU 12.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
fatal : ni ceci ni aucun de ses répertoires parents (jusqu'au point de montage /) n'est un dépôt git
Arrêt à la limite du système de fichiers (GIT_DISCOVERY_ACROSS_FILESYSTEM n'est pas défini).
CMake Warning at CMakeLists.txt:12 (find_package):
  By not providing "FindCPM.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "CPM", but
  CMake did not find one.

  Could not find a package configuration file provided by "CPM" with any of
  the following names:

    CPMConfig.cmake
    cpm-config.cmake

  Add the installation prefix of "CPM" to CMAKE_PREFIX_PATH or set "CPM_DIR"
  to a directory containing one of the above files.  If "CPM" provides a
  separate development package or SDK, be sure it has been installed.


-- CPM: Adding package cxxopts@2.2.1 (v2.2.1)
-- cxxopts version 2.2.0
-- Configuring done
-- Generating done

```

---

Now it seems to work :) 

Please have a try too.. just in case

---

All good now. For an initial integration I'm :+1: on this.